feat(fdc): Data Connect client sdk caching

aashishpatil-g opened the pull request:
FDC SDK Caching. 

For Googlers, check go/fdc-sdk-caching.

aashishpatil-g committed:
Initial Commit

aashishpatil-g committed:
SQLite implementation

# Conflicts:
#	packages/firebase_data_connect/firebase_data_connect/pubspec.yaml

# Conflicts:
#	packages/firebase_data_connect/firebase_data_connect/pubspec.yaml

# Conflicts:
#	packages/firebase_data_connect/firebase_data_connect/pubspec.yaml

aashishpatil-g committed:
Cleanup part 1

aashishpatil-g committed:
Cleanup and Reactivity Support

aashishpatil-g committed:
Add CacheStorage to public API

aashishpatil-g committed:
CacheSettings constructor feedback from Cynthia

aashishpatil-g committed:
Cleanup and some fixes

aashishpatil-g committed:
fix warning about unclosed sink

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/lib/src/core/ref.dart:
```diff
@@ -249,13 +365,17 @@ class MutationRef<Data, Variables> extends OperationRef<Data, Variables> {
+
+    return OperationResult(dataConnect, typedData, DataSource.server, this);
   }
 }
```
This is currently not supported.

aashishpatil-g committed:
Gemini code review feedback

aashishpatil-g committed:
Gemini feedback - queryId construction being brittle

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/lib/src/cache/sqlite_cache_provider.dart:
```diff
@@ -0,0 +1,130 @@
+
+  final String entityDataTable = 'entity_data';
+  final String resultTreeTable = 'query_results';
+  final String entityDataRefsTable = 'entity_data_query_refs';
```
removed

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/lib/src/cache/cache_manager.dart:
```diff
@@ -0,0 +1,160 @@
+
+    // we have a provider lets ensure its initialized
+    bool? initialized = await providerInitialization;
+    if (initialized == null || initialized == false) {
```
changed

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/lib/src/firebase_data_connect.dart:
```diff
@@ -148,28 +173,33 @@ class FirebaseDataConnect extends FirebasePluginPlatform {
 
     if (cachedInstances[app.name] != null &&
         cachedInstances[app.name]![connectorConfig.toJson()] != null) {
+      print('Returning cached FirebaseDataConnect instance');
```
removed

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/test/src/common/common_library_test.dart:
```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import 'dart:nativewrappers/_internal/vm/bin/vmservice_io.dart';
```
removed

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/lib/src/firebase_data_connect.dart:
```diff
@@ -164,12 +198,16 @@ class FirebaseDataConnect extends FirebasePluginPlatform {
     }
 
+    //TODO remove after testing since CS should be null by default
+    cacheSettings = cacheSettings ?? CacheSettings();
```
This is temporary till we test the cache. Will be removed.

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/lib/src/cache/sqlite_cache_provider.dart:
```diff
@@ -0,0 +1,127 @@
+  @override
+  void manageCacheSize() {
+    // TODO: implement manageCacheSize
+  }
```
Will be fixed in a later PR.

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/lib/src/cache/cache_data_types.dart:
```diff
@@ -0,0 +1,309 @@
+        scalarValues: scalars,
+        nestedObjects: objects,
+        nestedObjectLists: objLists);
+  }
```
Keeping this since its easier to step through and debug

aashishpatil-g replied:
Duplicate feedback.

aashishpatil-g committed:
Cleanup of some dead or unnecessary code

aashishpatil-g commented on packages/firebase_data_connect/firebase_data_connect/lib/src/cache/cache_data_types.dart:
```diff
@@ -0,0 +1,306 @@
+        Map<String, dynamic> objValue = value as Map<String, dynamic>;
+        EntityNode node = EntityNode.fromJson(objValue, cacheProvider);
+        objects?[key] = node;
+      });
```
Easier to debug.

aashishpatil-g committed:
Gemini review feedback

cynthiajoan commented on packages/firebase_data_connect/firebase_data_connect/lib/src/firebase_data_connect.dart:
Do we need clearCache() method?

cynthiajoan commented on packages/firebase_data_connect/firebase_data_connect/lib/firebase_data_connect.dart:
```diff
@@ -39,3 +39,5 @@ export 'src/optional.dart'
         listDeserializer,
         listSerializer;
 export 'src/timestamp.dart' show Timestamp;
+export 'src/cache/cache_data_types.dart'
```
I see in the api doc there's a class DataSource, is it being implemented or we decide don't need it?

aashishpatil-g committed:
Test to check maxAge conformance

aashishpatil-g committed:
Fix git action failures for analyze, test

Removed a test no longer valid

aashishpatil-g commented:
Rebased off main, force pushed

aashishpatil-g committed:
Fix test errors

Moved tests from transport to ref.dart since decode has been moved out of transport.

aashishpatil-g committed:
e2eListen - explicit server fetch.

aashishpatil-g committed:
format fix

aashishpatil-g committed:
version constrain crypto dependency

aashishpatil-g committed:
Fix e2e test for listen

aashishpatil-g committed:
Add liceses to generated mocks

maneesht commented on packages/firebase_data_connect/firebase_data_connect/lib/src/cache/result_tree_processor.dart:
```diff
@@ -0,0 +1,147 @@
+    return DehydrationResult(rootNode, impactedQueryIds);
+  }
+
+  EntityNode _dehydrateNode(String queryId, dynamic data,
```
I noticed that this is called `dehydrateResults` in iOS: https://github.com/firebase/data-connect-ios-sdk/blob/9005fa987850e013394df14539b0aa24101bf88f/Sources/Cache/ResultTreeProcessor.swift#L66

maneesht commented on packages/firebase_data_connect/firebase_data_connect/lib/src/cache/cache_manager.dart:
I think the file should be renamed to `cache.dart`
